File: assets/js/ELTT-launcher.js
// assets/js/ELTT-launcher.js

"use strict";

/*
 * Frontend-Launcher-/DEX-Logik für die ELTT-Blockchain.
 * Deterministisch, auditierbar, paritätisch, souveränitätsfokussiert.
 * Keine Owner-Adresse, keine Seeds, keine privaten Schlüssel im Code.
 * Keine eigene Blockchain, kein eigener STATE.
 * Alle Operationen laufen über ELTTBlockchain.withPersistentBlockchain.
 */

const ELTTLauncher = (() => {
    const {
        TokenKind,
        TxKind,
        TokenType,
        Wallet,
        Transaction,
        LiquidityPool,
        appendBlock,
        computeTxEnergy,
        createPool,
        getPool,
        buildSwapTx,
        buildAddLiquidityTx,
        buildRemoveLiquidityTx,
        withPersistentBlockchain,
    } = window.ELTTBlockchain;

    const { getCurrentWalletAddress } = window.ELTTWallet;

    /* -------------------------
       Pools (persistent)
    -------------------------- */

    function ensurePool(tokenXIndex, tokenYIndex, lpTokenIndex) {
        return withPersistentBlockchain((bc) => {
            for (let i = 0; i < bc.pools.length; i++) {
                const p = bc.pools[i];
                if (
                    p.tokenXIndex === tokenXIndex &&
                    p.tokenYIndex === tokenYIndex &&
                    p.lpTokenIndex === lpTokenIndex
                ) {
                    return { pool: p, index: i };
                }
            }
            const pool = createPool(bc, tokenXIndex, tokenYIndex, lpTokenIndex);
            return { pool, index: bc.pools.length - 1 };
        });
    }

    /* -------------------------
       Transaktions-Builder (deterministisch)
    -------------------------- */

    function buildDeterministicSwapTx(toAddress, amountIn, tokenInIndex, memo) {
        const fromAddress = getCurrentWalletAddress();
        if (!fromAddress) {
            throw new Error("No current wallet address set for swap.");
        }
        const tx = buildSwapTx(
            fromAddress,
            toAddress,
            amountIn,
            tokenInIndex,
            memo || ""
        );
        computeTxEnergy(tx);
        return tx;
    }

    function buildDeterministicAddLiquidityTx(poolAddress, amountX, tokenXIndex, memo) {
        const fromAddress = getCurrentWalletAddress();
        if (!fromAddress) {
            throw new Error("No current wallet address set for add-liquidity.");
        }
        const tx = buildAddLiquidityTx(
            fromAddress,
            poolAddress,
            amountX,
            tokenXIndex,
            memo || ""
        );
        computeTxEnergy(tx);
        return tx;
    }

    function buildDeterministicRemoveLiquidityTx(poolAddress, amountLp, lpTokenIndex, memo) {
        const fromAddress = getCurrentWalletAddress();
        if (!fromAddress) {
            throw new Error("No current wallet address set for remove-liquidity.");
        }
        const tx = buildRemoveLiquidityTx(
            fromAddress,
            poolAddress,
            amountLp,
            lpTokenIndex,
            memo || ""
        );
        computeTxEnergy(tx);
        return tx;
    }

    /* -------------------------
       Lokale DEX-Effekte (persistent im Engine-STATE)
    -------------------------- */

    function applyAddLiquidityLocally(poolIndex, amountX, amountY) {
        return withPersistentBlockchain((bc) => {
            const pool = getPool(bc, poolIndex);
            if (!pool) {
                return null;
            }
            pool.reserveX += amountX;
            pool.reserveY += amountY;
            return {
                tokenXIndex: pool.tokenXIndex,
                tokenYIndex: pool.tokenYIndex,
                lpTokenIndex: pool.lpTokenIndex,
                reserveX: pool.reserveX,
                reserveY: pool.reserveY,
            };
        });
    }

    function applyRemoveLiquidityLocally(poolIndex, shareX, shareY) {
        return withPersistentBlockchain((bc) => {
            const pool = getPool(bc, poolIndex);
            if (!pool) {
                return null;
            }
            pool.reserveX -= shareX;
            pool.reserveY -= shareY;
            if (pool.reserveX < 0.0) pool.reserveX = 0.0;
            if (pool.reserveY < 0.0) pool.reserveY = 0.0;
            return {
                tokenXIndex: pool.tokenXIndex,
                tokenYIndex: pool.tokenYIndex,
                lpTokenIndex: pool.lpTokenIndex,
                reserveX: pool.reserveX,
                reserveY: pool.reserveY,
            };
        });
    }

    /* -------------------------
       Commit (persistent)
    -------------------------- */

    function commitDexTransactions(transactions, timestamp) {
        return withPersistentBlockchain((bc) => {
            const txs = transactions.slice();
            for (const tx of txs) {
                if (!tx.energy || tx.energy === 0.0) {
                    computeTxEnergy(tx);
                }
            }
            const block = appendBlock(bc, txs, timestamp);
            return {
                index: block.index,
                prevHash: block.prevHash,
                timestamp: block.timestamp,
                hash: block.hash,
            };
        });
    }

    /* -------------------------
       Export (read-only Sicht)
    -------------------------- */

    function exportDexState() {
        return withPersistentBlockchain((bc) => {
            return {
                pools: bc.pools.map((p, idx) => ({
                    index: idx,
                    tokenXIndex: p.tokenXIndex,
                    tokenYIndex: p.tokenYIndex,
                    lpTokenIndex: p.lpTokenIndex,
                    reserveX: p.reserveX,
                    reserveY: p.reserveY,
                })),
                blocks: bc.blocks.map((b) => ({
                    index: b.index,
                    prevHash: b.prevHash,
                    timestamp: b.timestamp,
                    hash: b.hash,
                })),
            };
        });
    }

    return Object.freeze({
        ensurePool,
        buildDeterministicSwapTx,
        buildDeterministicAddLiquidityTx,
        buildDeterministicRemoveLiquidityTx,
        applyAddLiquidityLocally,
        applyRemoveLiquidityLocally,
        commitDexTransactions,
        exportDexState,
    });
})();

window.ELTTLauncher = ELTTLauncher;

File: PYTHON-ELTT-MODULE-PARITY/ELTT-liquidity-pools/ELTT-launcher.py
# PYTHON-ELTT-MODULE-PARITY/ELTT-liquidity-pools/ELTT-launcher.py
#
# Paritätsmodul für die Launcher-Sicht im Liquidity-Pool-Kontext.
# Spiegelt deterministisch die zentrale ELTT-Blockchain-Logik.
#
# Keine Beispiele, keine Interpretationen, keine Abweichungen.

from ELTT-Blockchain import (  # type: ignore
    Blockchain,
    LiquidityPool,
    Transaction,
    TokenType,
    TokenKind,
    TxKind,
    compute_tx_energy,
)


def create_pool(bc: Blockchain,
                token_x_index: int,
                token_y_index: int,
                lp_token_index: int) -> LiquidityPool:
    """
    Deterministische Erstellung eines Liquidity-Pools.
    """
    pool = LiquidityPool(
        token_x_index=token_x_index,
        token_y_index=token_y_index,
        reserve_x=0.0,
        reserve_y=0.0,
        lp_token_index=lp_token_index,
    )
    bc.pools.append(pool)
    return pool


def add_liquidity(bc: Blockchain,
                  pool_index: int,
                  amount_x: float,
                  amount_y: float) -> None:
    """
    Deterministisches Hinzufügen von Liquidität.
    """
    if pool_index < 0 or pool_index >= len(bc.pools):
        return
    pool = bc.pools[pool_index]
    pool.reserve_x += amount_x
    pool.reserve_y += amount_y


def remove_liquidity(bc: Blockchain,
                     pool_index: int,
                     share_x: float,
                     share_y: float) -> None:
    """
    Deterministisches Entfernen von Liquidität.
    """
    if pool_index < 0 or pool_index >= len(bc.pools):
        return
    pool = bc.pools[pool_index]
    pool.reserve_x -= share_x
    pool.reserve_y -= share_y
    if pool.reserve_x < 0.0:
        pool.reserve_x = 0.0
    if pool.reserve_y < 0.0:
        pool.reserve_y = 0.0


def build_add_liquidity_tx(from_address: str,
                           pool_address: str,
                           amount_x: float,
                           token_x_index: int,
                           memo: str = "") -> Transaction:
    """
    Deterministische Erstellung einer Add-Liquidity-Transaktion.
    """
    return Transaction(
        from_address=from_address,
        to_address=pool_address,
        amount=amount_x,
        token_index=token_x_index,
        kind=TxKind.ADD_LIQUIDITY,
        memo=memo[:128],
    )


def build_remove_liquidity_tx(from_address: str,
                              pool_address: str,
                              amount_lp: float,
                              lp_token_index: int,
                              memo: str = "") -> Transaction:
    """
    Deterministische Erstellung einer Remove-Liquidity-Transaktion.
    """
    return Transaction(
        from_address=from_address,
        to_address=pool_address,
        amount=amount_lp,
        token_index=lp_token_index,
        kind=TxKind.REMOVE_LIQUIDITY,
        memo=memo[:128],
    )


__all__ = [
    "Blockchain",
    "LiquidityPool",
    "Transaction",
    "TokenType",
    "TokenKind",
    "TxKind",
    "compute_tx_energy",
    "create_pool",
    "add_liquidity",
    "remove_liquidity",
    "build_add_liquidity_tx",
    "build_remove_liquidity_tx",
]
